pkg/geo/geomfn/subdivide.go

```diff
@@ -0,0 +1,364 @@
+// licenses/APL.txt.
+
+package geomfn
+
+import (
+	"math"
+
+	"github.com/cockroachdb/cockroach/pkg/geo"
+	"github.com/cockroachdb/errors"
+	"github.com/twpayne/go-geom"
+)
+
+// Subdivide returns a geometry divided into parts, where each part contains no more than the number of vertices provided.
+func Subdivide(g geo.Geometry, maxVertices int) (geo.Geometry, error) {
+	if g.Empty() {
+		return g, nil
+	}
+	const minMaxVertices = 5
+	if maxVertices < minMaxVertices {
+		return geo.Geometry{}, errors.Newf("max_vertices number cannot be less than %v", minMaxVertices)
+	}
+
+	gt, err := g.AsGeomT()
+	if err != nil {
+		return geo.Geometry{}, errors.Newf("could not transform input geometry into geom.T: %v", err)
+	}
+	col := geom.NewGeometryCollection()
+	dim, err := dimensionFromGeomT(gt)
+	if err != nil {
+		return geo.Geometry{}, errors.Newf("could not calculate geometry dimension: %v", err)
+	}
+	const startDepth = 0
+	err = subdivideRecursive(gt, maxVertices, startDepth, col, dim)
+	if err != nil {
+		return geo.Geometry{}, err
+	}
+	col.SetSRID(gt.SRID())
+	output, err := geo.MakeGeometryFromGeomT(col)
+	if err != nil {
+		return geo.Geometry{}, errors.Newf("could not transform output geom.T into geometry: %v", err)
+	}
+	return output, nil
+}
+
+func subdivideRecursive(
+	gt geom.T, maxVertices int, depth int, col *geom.GeometryCollection, dim int,
+) error {
+	const geometryCollectionPushError = "could not add geometry to results collection: %v"
+	// maxDepth 50 => 2^50 ~= 10^15 subdivisions.
+	const maxDepth = 50
+	clip := geo.BoundingBoxFromGeomTGeometryType(gt)
+	width := clip.HiX - clip.LoX
+	height := clip.HiY - clip.LoY
+	if width == 0 && height == 0 {
+		if gt, ok := gt.(*geom.Point); ok && dim == 0 {
+			err := col.Push(gt)
+			if err != nil {
+				return errors.Newf(geometryCollectionPushError, err)
+			}
+		}
+		return nil
+	}
+	// Boundaries needs to be widen slightly for intersection to work properly.
+	tolerance := 1e-12
+	if width == 0 {
+		clip.HiX += tolerance
+		clip.LoX -= tolerance
+		width = 2 * tolerance
+	}
+	if height == 0 {
+		clip.HiY += tolerance
+		clip.LoY -= tolerance
+		height = 2 * tolerance
+	}
+	// Recurse into collections other than MultiPoint.
+	isCollection, err := isGeomTCollection(gt)
+	if err != nil {
+		return errors.Newf("error checking if geom.T is collection: %v", err)
+	}
+	_, isMultiPoint := gt.(*geom.MultiPoint)
+	if isCollection && !isMultiPoint {
+		err := forEachGeomTInCollection(gt, func(gi geom.T) error {
+			// It's not a subdivision yet, so depth stays the same
+			return subdivideRecursive(gi, maxVertices, depth, col, dim)
+		})
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+	currDim, err := dimensionFromGeomT(gt)
+	if err != nil {
+		return errors.Newf("error checking geom.T dimension: %v", err)
+	}
+	if currDim < dim {
+		// Ignore lower dimension object produced by clipping at a shallower recursion level.
+		return nil
+	}
+	if depth > maxDepth {
```

Comment:
Please note that this behaviour is not covered in the tests - testing it as it is will slow down the unit tests, while mocking the constant value requires a small workaround and I was not sure if it's worth it.

Reply:
i think you still can!
if you make this function signature
```

func subdivideRecursive(
	gt geom.T, maxVertices int, depth int, col *geom.GeometryCollection, dim int, maxDepth int,
) error {
```

and use a different value for maxDepth, as well as use a constant for the initial call, that avoids the mocking issue!

Reply:
That's a good idea, fixed 👍